src/compiler/checker.ts

```diff
@@ -11133,7 +11133,53 @@ namespace ts {
             return -1;
         }
 
-        function isInLegalParameterTypePredicatePosition(node: Node): boolean {
+        function checkTypePredicate(node: TypePredicateNode) {
+            const parent = getTypePredicateParent(node);
+            if (!parent) {
+                return;
+            }
+            const returnType = getReturnTypeOfSignature(getSignatureFromDeclaration(parent));
+            if (!returnType || !(returnType.flags & TypeFlags.PredicateType)) {
+                return;
+            }
+            const { parameterName } = node;
+            if (parameterName.kind === SyntaxKind.ThisType) {
+                getTypeFromThisTypeNode(parameterName as ThisTypeNode);
+            }
+            else {
+                const typePredicate = <IdentifierTypePredicate>(<PredicateType>returnType).predicate;
+                if (typePredicate.parameterIndex >= 0) {
+                    if (parent.parameters[typePredicate.parameterIndex].dotDotDotToken) {
+                        error(parameterName,
+                            Diagnostics.A_type_predicate_cannot_reference_a_rest_parameter);
+                    }
+                    else {
+                        checkTypeAssignableTo(typePredicate.type,
+                            getTypeOfNode(parent.parameters[typePredicate.parameterIndex]),
+                            node.type);
+                    }
+                }
+                else if (parameterName) {
+                    let hasReportedError = false;
+                    for (const { name } of parent.parameters) {

```

Comment:
While writing up the new issue, I realised that this whole branch is an error case because `parameterIndex` is unknown (`<0` actually). `getTypePredicateParameterIndex` couldn't find it in the top-level so the applicable declaration is in turn unknown. It falls to `checkIfTypePredicateVariableIsDeclaredInBindingPattern` to retrieve the erroneously nested declaration.

In this example, `a` is illegal because it's found in a binding pattern:
```ts
declare f([nested, b]: [any, number], [unused, irrelevant]: [string, string]): nested is number;
```